Allow disjoint mutable world access via `EntityMut`

JoJoJet opened the pull request:
# Objective

Fix #4278
Fix #5504
Fix #9422

Provide safe ways to borrow an entire entity, while allowing disjoint mutable access.  `EntityRef` and `EntityMut` are not suitable for this, since they provide access to the entire world -- they are just helper types for working with `&World`/`&mut World`.

This has potential uses for reflection and serialization

## Solution

Remove `EntityRef::world`, which allows it to soundly be used within queries.

`EntityMut` no longer supports structural world mutations, which allows multiple instances of it to exist for different entities at once. Structural world mutations are performed using the new type `EntityWorldMut`.

```rust
fn disjoint_system(
     q2: Query<&mut A>,
     q1: Query<EntityMut, Without<A>>,
) { ... }

let [entity1, entity2] = world.many_entities_mut([id1, id2]);
*entity1.get_mut::<T>().unwrap() = *entity2.get().unwrap();

for entity in world.iter_entities_mut() {
    ...
}
```

---

## Changelog

- Removed `EntityRef::world`, to fix a soundness issue with queries.
+ Removed the ability to structurally mutate the world using `EntityMut`, which allows it to be used in queries.
+ Added `EntityWorldMut`, which is used to perform structural mutations that are no longer allowed using `EntityMut`.

## Migration Guide

**Note for maintainers: ensure that the guide for #9604 is updated accordingly.**

Removed the method `EntityRef::world`, to fix a soundness issue with queries. If you need access to `&World` while using an `EntityRef`, consider passing the world as a separate parameter.

`EntityMut` can no longer perform 'structural' world mutations, such as adding or removing components, or despawning the entity. Additionally, `EntityMut::world`, `EntityMut::world_mut` , and `EntityMut::world_scope` have been removed.
Instead, use the newly-added type `EntityWorldMut`, which is a helper type for working with `&mut World`.

alice-i-cecile commented on crates/bevy_ecs/src/query/access.rs:
```diff
@@ -113,39 +116,58 @@ impl<T: SparseSetIndex> Access<T> {
+        self.reads_all || self.writes_all
+    }
+
+    /// Returns `true` if this has write access to all indexec elements (i.e `EntityBorrowMut`).
```
```suggestion
    /// Returns `true` if this has write access to all indexed elements (i.e `EntityBorrowMut`).
```

alice-i-cecile reviewed (commented):
Part of me feels that `EntityRef` and `EntityMut` should be removed completely in this PR. This feels like a much more helpful abstraction that serves a very similar role.

Why is that a bad idea?

alice-i-cecile commented on crates/bevy_ecs/src/world/entity_borrow.rs:
```diff
@@ -0,0 +1,447 @@
+
+    #[test]
+    #[should_panic]
+    fn borrow_mut_incompatible1() {
```
These tests should have better names, even if they're more verbose. Maybe `borrow_mut_incompatible_with_read_only_component` and so on?

alice-i-cecile commented on crates/bevy_ecs/src/world/mod.rs:
```diff
@@ -385,12 +508,38 @@ impl World {
         })
     }
 
+    /// Returns a mutable iterator over current entities.
```
```suggestion
    /// Returns a mutable iterator over all entities in the `World`.
```

alice-i-cecile commented on crates/bevy_ecs/src/world/mod.rs:
```diff
@@ -385,12 +508,38 @@ impl World {
     }
 
+    /// Returns a mutable iterator over current entities.
+    pub fn iter_entities_mut(&mut self) -> impl Iterator<Item = EntityBorrowMut<'_>> + '_ {
```
Should we have an immutable equivalent?

JoJoJet replied:
There already is one :)

alice-i-cecile commented:
Generally impressed: this is an extremely useful bit of API (especially for reflection-adjacent tasks) and is well-made. The question about `EntityRef` and `EntityMut` is my main concern: they seem rather duplicative and less elegant.

If we want to keep them, I'd be interested in a way to go *back* to the `World`-containing variants, by passing in `&World` or `&mut World` respectively.

JoJoJet commented:
> Part of me feels that EntityRef and EntityMut should be removed completely in this PR. This feels like a much more helpful abstraction that serves a very similar role.
> Why is that a bad idea?

`EntityMut` is still useful on its own. It allows you to insert and remove components, which `EntityBorrowMut` cannot do.

I could see an argument for removing `EntityRef` though. I mostly left it because `EntityRef::world` is mildly useful sometimes.

JoJoJet commented:
I've updated this to remove the `WorldQuery` impl for `EntityRef`, in order to fix #9422.

alice-i-cecile reviewed (approved):
Right: `EntityMut` is genuinely useful. Alright, let's move forward like this for now.

iiYese commented on crates/bevy_ecs/src/world/entity_borrow.rs:
```diff
@@ -0,0 +1,447 @@
+}
+
+#[cfg(test)]
+mod tests {
```
Shouldn't there also be a panicing test for `Query<EntityBorrow>` + `Query<&mut A>`?

iiYese commented:
> I mostly left it because `EntityRef::world` is mildly useful sometimes

I would like identified cases of when. The API duplication is getting ridiculous now
- `&mut World`
- `EntityRef`
- `EntityMut`
- `EntityBorrow`
- `EntityBorrowMut`

Should try not to bloat this further until refactors on the horizon are possible. If no good argument can be made for or against keeping `EntityRef` then a push for its removal should be made to find one.

JoJoJet commented:
> > I mostly left it because `EntityRef::world` is mildly useful sometimes
> 
> I would like identified cases of when.

I like to think of `EntityRef` as a performance-optimized `(Entity, &World)` tuple, which can be useful any time you're working with direct world access (I have used it for this before). If we removed `EntityRef::world`, then a bunch of users would have to change some method signatures from `fn(EntityRef)` to `fn(EntityRef, &World)` -- not the biggest loss but still annoying.

That said, I'm open to doing it if we decide it's a good idea. I think the path with the least API breakages would be to remove `EntityRef`, and rename `EntityBorrow` ->  `EntityRef`. This would open the question of what to call `EntityBorrowMut`, which would seem out of place without a corresponding `EntityBorrow`.

JoJoJet commented:
I've added some unit tests that are known to fail due to limitations in our current access model. Since the issue is with overly restrictive accesses, it will not cause undefined behavior, so I think it is okay to fix this issue later. Also, note that this issue already existed on the main branch with `EntityRef` queries.

JoJoJet commented:
I have merged `EntityBorrow` into `EntityRef`, and removed `EntityRef::world`.

JoJoJet commented:
I have also renamed `EntityMut` to `EntityWorldMut`, and renamed `EntityBorrowMut` to `EntityMut`. Since I've made so many changes, I've updated the PR description to give an overview of the current state of this PR.

JoJoJet commented on crates/bevy_ecs/src/world/entity_ref.rs:
```diff
@@ -166,35 +149,263 @@ impl<'w> EntityRef<'w> {
+/// # use bevy_ecs::prelude::*;
+/// # #[derive(Component)] pub struct A;
+/// fn disjoint_system(
+///     query1: Query<EntityRef, With<A>>,
+///     query2: Query<EntityRef, Without<A>>,
```
```suggestion
///     query1: Query<EntityMut, With<A>>,
///     query2: Query<EntityMut, Without<A>>,
```

JoJoJet commented on crates/bevy_ecs/src/world/entity_ref.rs:
```diff
@@ -1369,4 +1576,113 @@ mod tests {
+
+    #[test]
+    fn disjoint_access() {
+        fn disjoint_readonly(_: Query<EntityRef, With<A>>, _: Query<EntityRef, Without<A>>) {}
```
```suggestion
        fn disjoint_readonly(_: Query<EntityMut, With<A>>, _: Query<EntityRef, Without<A>>) {}
```

maniwani reviewed (approved):
This looks good to me. I mostly wanted to make sure the access is sound. 

I think renaming the existing types is a little strange, but ig it's better for queries to get the types with shorter names. It's fine if there's some type duplication between "queried" and "direct-world" access though. We may find a need for an `EntityWorldRef` later.

alice-i-cecile commented:
I think there are subtle merge conflicts here @JoJoJet
